README


This archive contains the distribution Mixin-Historian,
version 0.001:

  a mixin for recording history about objects

This software is copyright (c) 2013 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Mixin-Historian

0.001     2013-10-05 07:54:44 America/New_York
          update packaging

          write some documentation, for Pete's sake

0.101350  2010-05-15 14:39:41 America/New_York
					first CPAN release


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "a mixin for recording history about objects",
  "AUTHOR" => "Ricardo Signes <rjbs\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Mixin-Historian",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Mixin::Historian",
  "PREREQ_PM" => {
    "Mixin::ExtraFields" => "0.008",
    "Sub::Exporter" => 0,
    "Sub::Exporter::ForMethods" => 0,
    "Test::Deep" => 0,
    "Test::More" => "0.88",
    "base" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::Deep" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Mixin/Historian/Driver.pm
use strict;
use warnings;
package Mixin::Historian::Driver;
{
  $Mixin::Historian::Driver::VERSION = '0.001';
}
# ABSTRACT: base class for classes that act as Historian storage drivers


1;

__END__

=pod

=head1 NAME

Mixin::Historian::Driver - base class for classes that act as Historian storage drivers

=head1 VERSION

version 0.001

=head1 METHODS

Classes extending Mixin::Historian::Driver are expected to provide the
following methods:

=head1 new

This method gets the driver configuration from the call to
L<Mixin::Historian>'s C<import> method and should return a new driver instance.

=head1 add_history

This method is passed an arrayref of the argument(s) to the generated and
installed C<add_history> method.  It is is expected to store the history entry.

=head1 AUTHOR

Ricardo Signes <rjbs@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Mixin/Historian/Driver/Array.pm
use strict;
use warnings;
package Mixin::Historian::Driver::Array;
{
  $Mixin::Historian::Driver::Array::VERSION = '0.001';
}
use base 'Mixin::Historian::Driver';
# ABSTRACT: a driver that stores history in an in-memory array (for testing)


sub new {
  my ($class, $arg) = @_;

  return bless {
    array => [],
  } => $class;
}

sub _array {
  $_[0]{array};
}

sub entries {
  my ($self) = @_;
  return @{ $self->_array };
}

sub add_history {
  my ($self, $arg) = @_;

  my $record = $arg->{args}[0];

  push @{ $self->_array }, {
    time   => time,
    record => $record,
  };
}

1;

__END__

=pod

=head1 NAME

Mixin::Historian::Driver::Array - a driver that stores history in an in-memory array (for testing)

=head1 VERSION

version 0.001

=head1 DESCRIPTION

This driver, meant primarily for testing, logs history events as hashrefs in an
in-memory arrayref stored in the driver.

The events may accessed by the driver's C<entries> method, and are returned as
a list of hashrefs in the form:

  {
    time   => $epoch_seconds,
    record => $hashref_passed_to_add_history,
  }

=head1 AUTHOR

Ricardo Signes <rjbs@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Mixin/Historian.pm
use strict;
use warnings;
package Mixin::Historian;
{
  $Mixin::Historian::VERSION = '0.001';
}
use Mixin::ExtraFields 0.008 ();
use base 'Mixin::ExtraFields';
# ABSTRACT: a mixin for recording history about objects

use Sub::Exporter::ForMethods ();

use Sub::Exporter -setup => {
  groups => {
    history => \'gen_fields_group',
  },
  installer => Sub::Exporter::ForMethods::method_installer(),
};


sub default_moniker { 'history' }

sub driver_base_class { 'Mixin::Historian::Driver' }

sub methods { qw(add) }

sub driver_method_name {
  my ($self, $method) = @_;
  $self->method_name($method, 'history');
}

sub build_method {
  my ($self, $method_name, $arg) = @_;

  # Remember that these are all passed in as references, to avoid unneeded
  # copying. -- rjbs, 2006-12-07
  my $id_method = $arg->{id_method};
  my $driver    = $arg->{driver};

  my $driver_method  = $self->driver_method_name($method_name);

  return sub {
    my $object = shift;
    my $id     = $object->$$id_method;
    Carp::confess "couldn't determine id for object" unless defined $id;
    $$driver->$driver_method({
      object => $object,
      mixin  => $self,
      id     => $id,
      args   => \@_,
    });
  };
}

1;

__END__

=pod

=head1 NAME

Mixin::Historian - a mixin for recording history about objects

=head1 VERSION

version 0.001

=head1 SYNOPSIS

  package My::Object;
  use Mixin::Historian -history => {
    driver => {
      class => 'YourDriver',
      ...,
    },
  };

  # Later...
  my $object = My::Object->retrieve(1234);

  $object->add_history({
    type     => 'lava damage',
    severity => 'very badly burned',
    volcano  => 'Eyjafjallajokull',
  });

=head1 DESCRIPTION

Mixin::Historian is an application of Mixin::ExtraFields.  If you're not
familiar with it, you should read about it, both in L<its
documentation|Mixin::ExtraFields> and in L<this article about
Mixin::ExtraFields|http://advent.rjbs.manxome.org/2009-12-22.html>.

Generally, it provides simple mechanism for write-only history.  Importing the
C<-history> group will get you the C<add_history> method, which generally will
accept one hashref with at least a C<type> key.  This will be passed along to
the driver's C<add_history> method.

=head1 TODO

I have shoehorned an extra layer of functionality into the Historian driver
that I use in my employer's code.  When initialized, the Historian mixin is
told all legal types, something like this:

  type_map => {
    'lava damage' => {
      severity => { required => 1, store_as => 'extra_1' },
      volcano  => { required => 0, store_as => 'extra_2' },
    },
    ...
  }

This way, history entries can be validated before writing.  The C<store_as>
entries indicate how the arguments to C<add_history> are mapped to database
columns.  The entire argument is also stored in one field as JSON, and a few
other attributes are always required (like C<by_whom>) and some are added just
in time (like C<logged_at>).

This feature is not yet present in the CPAN library because I have not yet
found a suitable decomposition of concerns to make it a component.

=head1 AUTHOR

Ricardo Signes <rjbs@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
